### ISSUE TITLE
Help: Interprocedural flow analysis and parameter tracking -- how does it work

### ISSUE THREAD

**Issue by nschuessler** [dotnet/roslyn-analyzers#5990]
I have implemented a flow analysis to detect duplicate dictionary accesses.  Interprocedural analysis is enabled. On the following code, it is flagging this as a duplicate analysis even though it is plainly not:

```csharp
public void DoSomething(IDictionary<string, string> dict)
{
     string a = Extract("keyA");
     string b = Extract("keyB");
     string c = Extract("keyC");

     string Extract(string key) => dict.TryGetValue(key, out string v) ? v : null;
}
```

In debugging through the operation visitor implementation, it seems that it is evaluting it as if it was written like this:

```csharp
public void DoSomething(IDictionary<string, string> dict)
{
     string a = dict.TryGetValue(key, out string v) ? v : nul;;
     string b = dict.TryGetValue(key, out string v) ? v : nul);
     string c = dict.TryGetValue(key, out string v) ? v : nul);

     string Extract(string key) => dict.TryGetValue(key, out string v) ? v : null;
}
```

which would trigger the rule.

The bulk of our analysis happens in the operation vistior overload `VisitInvocation_NonLambdaOrDelegateOrLocalFunction`.. In that function they parameter is given as `key` as you see in the second code snippet.

I assume there is a problem with our implementations of `SetValueForParameterOnEntry` and potentially `StopTrackingDataForParameter` overloads which is just cut/pasted code from another example.  Can you clarify what these methods are supposed to do and when they are called. If this is not potentially the issue, what else could be the problem?

Also, would your suggestion handle the next obvious case:

```csharp

public void DoSomething(IDictionary<string, string> dict)
{
    foreach (string myKey in new string[] { "KeyA", "KeyB", "KeyC" })
    {
          string a = Extract(myKey);
    }

     string Extract(string key) => dict.TryGetValue(key, out string v) ? v : null;
}
```

Thanks

**Comment by nschuessler** [comment]
@mavasani Maybe the person who knows the answer to this.
